module/data/item/discipline.mjs

```diff
@@ -1,41 +1,341 @@
+      }
+    );
+    return [
+      {
+        name:      ED4E.circleTalentRequirements[
+          game.settings.get( "ed4e", "lpTrackingCircleTalentRequirements" )
+        ],
+        value:     requirementValue,
+        fulfilled,
+      }
+    ];
+  }
+
+  get _talentRequirementsOptional() {
+    const nextLevel = this.level + 1;
+    const allCorrespondingTalents = this.talentsFromDiscipline;
+    const hasTalentFromCurrentCircle = allCorrespondingTalents.some(
+      talent => talent.system.source.atLevel === this.level
+    );
+    const talentsOnMinRank = allCorrespondingTalents.filter(
+      talent => talent.system.level === nextLevel <= 11 ? nextLevel : nextLevel - 1
+    );
+    const numTalentsOnMinRank = talentsOnMinRank.length;
+    const numTalentsRequired = nextLevel + 3;
+    const hasEnoughTalents = numTalentsOnMinRank >= numTalentsRequired;
+
+    return [
+      {
+        name:      "ED.Dialogs.Legend.Validation.hasTalentFromCurrentCircle",
+        value:     game.i18n.format(
+          hasTalentFromCurrentCircle
+            ? "ED.Dialogs.Legend.Validation.talentFromCurrentCircleOnNewLevel"
+            : "ED.Dialogs.Legend.Validation.noTalentFromCurrentCircleOnNewLevel",
+          { nextLevel },
+        ),
+        fulfilled: hasTalentFromCurrentCircle,
+      },
+      {
+        name:      "ED.Dialogs.Legend.Validation.hasTalentsOnMinRank",
+        value:     game.i18n.format(
+          hasEnoughTalents
+            ? "ED.Dialogs.Legend.Validation.enoughTalentsOnMinRank"
+            : "ED.Dialogs.Legend.Validation.notEnoughTalentsOnMinRank",
+          {
+            required: numTalentsRequired,
+            current:  numTalentsOnMinRank,
+            nextLevel,
+          }
+        ),
+        fulfilled: hasEnoughTalents,
+      },
+    ];
+  }
+
+  get _talentRequirementsHouseRule() {
+    const nextLevel = this.level + 1;
+    const tierInfos = {
+      novice:     this.getTalentsByTier( "novice" ),
+      journeyman: this.getTalentsByTier( "journeyman" ),
+      warden:     this.getTalentsByTier( "warden" ),
+      master:     this.getTalentsByTier( "master" ),
+    };
+    const minTalentsPerTier = {
+      novice:     nextLevel + 4,
+      journeyman: nextLevel >= 6 ? nextLevel - 5 : 0,
+      warden:     nextLevel >= 10 ? nextLevel - 9 : 0,
+      master:     nextLevel >= 14 ? nextLevel - 13 : 0,
+    };
 
-    /** @inheritDoc */
-    static migrateData( source ) {
-        super.migrateData( source );
-        // specific migration functions
+    const hasRequiredNoviceTalents = tierInfos.novice.length >= minTalentsPerTier.novice;
+    const hasRequiredJourneymanTalents = tierInfos.journeyman.length >= minTalentsPerTier.journeyman;
+    const hasRequiredWardenTalents = tierInfos.warden.length >= minTalentsPerTier.warden;
+    const hasRequiredMasterTalents = tierInfos.master.length >= minTalentsPerTier.master;
+
+    const hasTalentFromCurrentCircle = this.talentsFromDiscipline.some(
```

Comment:
The idea is to have at least 4+circle talents at the nextLevel rank. it seems not to be covered by this